File: 04 - Largest Palindrome Product/cpp/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int reverseNumber(int num)
{
    int reversed = 0;
    while (num > 0)
    {
        reversed = (reversed * 10) + (num % 10);
        num = num / 10;
    }
    return reversed;
}

bool isPalindrome(int num)
{
    return num == reverseNumber(num);
}

int largestPalindromeProduct(int low, int high)
{
    std::vector<int> palindromes = {};
    for (int i = low; i < high; i++)
    {
        for (int j = low; j < high; j++)
        {
            int product = i * j;
            if (isPalindrome(product))
            {
                palindromes.push_back(product);
            }
        }
    }
    std::sort(palindromes.begin(), palindromes.end());
    return palindromes.back();
}

int main()
{
    int result = largestPalindromeProduct(100, 1000);
    std::cout << "result: " << result << std::endl;
}

File: 04 - Largest Palindrome Product/nim/main.nim
import std/algorithm

proc reverse_number(num: int): int =
    var num = num
    var reversed: int = 0
    while num > 0:
        reversed = (reversed * 10) + (num mod 10)
        num = num div 10
    reversed

proc is_palindrome(num: int): bool =
    num == reverse_number(num)

proc largest_palindrome_product(low: int, high: int): int =
    var palindromes = newSeq[int]()
    for i in low ..< high:
        for j in low ..< high:
            let product: int = i * j
            if is_palindrome(product):
                palindromes.add(product)
    palindromes.sort()
    palindromes[^1]

let result: int = largest_palindrome_product(100, 1000)
echo "result: ", result
